modules/core/src/smithy4s/schema/SchemaPartition.scala

```diff
@@ -0,0 +1,192 @@
+/*
+ *  Copyright 2021-2022 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s.schema
+
+import smithy4s.PartialData
+import smithy4s.kinds.PolyFunction
+import Schema._
+
+/**
+  * A structure indicating the match result of running `Schema#partition` against a given predicate
+  *
+  *   - if the schema is not of a structure, or if none of the fields matched, then `NoMatch` should be returned
+  *   - if the schema is a structure and only a subset of its fields pass the predicate, then `PartialMatch` should be returned
+  *   - if the schema is a structure and all of its fields pass the predicate, then `TotalMatch` should be returned
+  */
+sealed trait SchemaPartition[A]
+
+object SchemaPartition {
+
+  // format: off
+  final case class TotalMatch[A](schema: Schema[A])                                                         extends SchemaPartition[A]
+  final case class SplittingMatch[A](matching: Schema[PartialData[A]], notMatching: Schema[PartialData[A]]) extends SchemaPartition[A]
+  final case class NoMatch[A]()                                                                             extends SchemaPartition[A]
```

Comment:
**thought:** not sure about how/if the call sites would be interested, but shouldn't this wrap a Schema[A] too?

Reply:
Probably not : the reason the `Total` wraps it is that the schema it wraps might not actually be the same as the input schema that gets partitioned.

Reply:
ohh right

Reply:
maybe worth a code comment on these classes to highlight the difference?

Reply:
https://github.com/disneystreaming/smithy4s/pull/875/commits/5f310af64819b41a0d4c6d5b104abecceb397520